pubsub/natspubsub/nats_test.go

```diff
@@ -0,0 +1,302 @@
+	"context"
+	"fmt"
+	"testing"
+
+	"gocloud.dev/gcerrors"
+	"gocloud.dev/pubsub"
+	"gocloud.dev/pubsub/driver"
+	"gocloud.dev/pubsub/drivertest"
+
+	"github.com/nats-io/gnatsd/server"
+	gnatsd "github.com/nats-io/gnatsd/test"
+	"github.com/nats-io/go-nats"
+)
+
+const (
+	TEST_PORT  = 11222
+	BENCH_PORT = 9222
+)
+
+type harness struct {
+	s  *server.Server
+	nc *nats.Conn
+}
+
+func newHarness(ctx context.Context, t *testing.T) (drivertest.Harness, error) {
+	opts := gnatsd.DefaultTestOptions
+	opts.Port = TEST_PORT
+	s := gnatsd.RunServer(&opts)
+	nc, err := nats.Connect(fmt.Sprintf("nats://127.0.0.1:%d", TEST_PORT))
+	if err != nil {
+		return nil, err
+	}
+	return &harness{s, nc}, nil
+}
+
+func (h *harness) CreateTopic(ctx context.Context, testName string) (driver.Topic, func(), error) {
+	cleanup := func() {}
+	dt := createTopic(h.nc, testName)
+	return dt, cleanup, nil
+}
+
+func (h *harness) MakeNonexistentTopic(ctx context.Context) (driver.Topic, error) {
+	// A nil *topic behaves like a nonexistent topic.
+	return (*topic)(nil), nil
+}
+
+func (h *harness) CreateSubscription(ctx context.Context, dt driver.Topic, testName string) (driver.Subscription, func(), error) {
+	ds := createSubscription(h.nc, testName)
+	// FIXME(dlc) - Check for error?
+	cleanup := func() {
+		var sub *nats.Subscription
+		if ds.As(&sub) {
+			sub.Unsubscribe()
+		}
+	}
+	return ds, cleanup, nil
+}
+
+func (h *harness) MakeNonexistentSubscription(ctx context.Context) (driver.Subscription, error) {
+	return (*subscription)(nil), nil
+}
+
+func (h *harness) Close() {
+	h.nc.Close()
+	h.s.Shutdown()
+}
+
+type natsAsTest struct{}
+
+func (natsAsTest) Name() string {
+	return "nats test"
+}
+
+func (natsAsTest) TopicCheck(top *pubsub.Topic) error {
+	var c2 nats.Conn
+	if top.As(&c2) {
+		return fmt.Errorf("cast succeeded for %T, want failure", &c2)
+	}
+	var c3 *nats.Conn
+	if !top.As(&c3) {
+		return fmt.Errorf("cast failed for %T", &c3)
+	}
+	return nil
+}
+
+func (natsAsTest) SubscriptionCheck(sub *pubsub.Subscription) error {
+	var c2 nats.Subscription
+	if sub.As(&c2) {
+		return fmt.Errorf("cast succeeded for %T, want failure", &c2)
+	}
+	var c3 *nats.Subscription
+	if !sub.As(&c3) {
+		return fmt.Errorf("cast failed for %T", &c3)
+	}
+	return nil
+}
+
+func (natsAsTest) TopicErrorCheck(t *pubsub.Topic, err error) error {
+	return nil
+}
+
+func (natsAsTest) SubscriptionErrorCheck(s *pubsub.Subscription, err error) error {
+	return nil
+}
+
+func (r natsAsTest) MessageCheck(m *pubsub.Message) error {
+	var pm nats.Msg
+	if m.As(&pm) {
+		return fmt.Errorf("cast succeeded for %T, want failure", &pm)
+	}
+	var ppm *nats.Msg
+	if !m.As(&ppm) {
+		return fmt.Errorf("cast failed for %T", &ppm)
+	}
+	return nil
+}
+
+func TestConformance(t *testing.T) {
+	asTests := []drivertest.AsTest{natsAsTest{}}
+	drivertest.RunConformanceTests(t, newHarness, asTests, nil)
+}
+
+// These are natspubsub specific to increase coverage.
+func TestSimplePubSub(t *testing.T) {
```

Comment:
Does this test actually add coverage? It looks like functionality that should be covered by the conformance tests.

Reply:
It did, not sure if you changed tests but when I added them yes they increased coverage.